### PR TITLE
feat: `EmbeddingOutput` to allow for custom output key and transformations

### PR DESCRIPTION
**Pull request by denwong47**
This PR enables a _slightly_ more structured way to handle the output types available in ONNX models. Currently, once the session had generated `SessionOutputs`, [`TextEmbedding::embed`](https://github.com/Anush008/fastembed-rs/blob/dd3a7af42d6ab4e152411abd023093e92009ddab/src/text_embedding.rs#L285) will either

- pick the only key if that's what the model output contains; or
- try the hardcoded `"last_hidden_state"` key, and panics if it doesn't exist without a specific explanation.

While output types like `"last_hidden_state"` and `"pooler_output"` are commonly supported, the actual keys exported by the model is controlled by ONNX configuration; `"last_hidden_state"` is not guaranteed to exist.

Once the output had been fetched with the key, `TextEmbedding::embed` will assume the output matrix is 3-dimentional; it will panic otherwise due to a `ArrayBase::slice` using 3 axis.

Apparently this may not always hold true - and for text embeddings 2-dimensional (No of input x Embeddings Dim) could perhaps be more common.

This PR introduces a check and deals with 2D/3D matrix differently.

### Motivation

This came about when I tried to use [`sentence-transformers/all-mpnet-base-v2`](https://huggingface.co/sentence-transformers/all-mpnet-base-v2) with `fastembed-rs`. The model does not natively have a ONNX equivalent, so editing `models_list()` to add the entry is not an option.

For this reason, `optimum-cli` was used to convert the model, and outputs were successfully generated from a `TextEmbedding<UserDefinedEmbeddingModel>` - only for the pipeline to panic afterwards.

I am assuming that the use of such a model is supported (at least evidently, it works); happy to be advised otherwise.

### Test Plan

An additional integration test had been added to `tests` folder behind an empty flag called `optimum-cli`. The test will externally call `optimum-cli` to pull models into the default cache folder. It is assumed that not every one will have the CLI tool or even Python installed, thus the flag is added to not run the test by default.

Once the models are pulled, they will be loaded as `UserDefinedEmbeddingModel` instances and dimension tested for their output.

If you have `optimum-cli` installed and available in your shell environment, you can run this test via

```sh
cargo test --features=optimum-cli
```

### Known Issues

Personally I would prefer the `output_key` to be exposed; these models from `sentence-transformers` appear to contain two outputs, `sentence_embedding` and `token_embeddings`; and perhaps there are endless variations out there we won't be able to capture. However as I could not introduce a new parameter to the function without breaking changes, I will leave this as a proposal here for maintainers to consider.

Many thanks for reviewing and much appreciate your time.

### CODE REVIEW

File: tests/optimum_cli_export.rs
Comment (Anush008) [inline tests/optimum_cli_export.rs:59]: This might become
```suggestion
        onnx_file: load_bytes_from_file(&output.join("model.onnx"))?,
        tokenizer_file: load_bytes_from_file(&output.join("tokenizer.json"))?
```
once #101 is merged and this PR is rebased.

File: Cargo.toml
Comment (Anush008) [inline Cargo.toml:41]: How about installing it in the the GitHub CI action too?

**Review by Anush008 — COMMENTED** [review]
Hey @denwong47. This is fantastic. Thanks a lot. The implementation and tests look great to me.

> However as I could not introduce a new parameter to the function without breaking changes,

We'll do a major release to introduce this PR, #101 and #99. So breaking changes are allowed.

**Comment by Anush008** [comment]
Also, please feel free to add yourself to the authors list in the `Cargo.toml` file.

**Comment by denwong47** [comment]
Thanks @Anush008, I'll add Python and `pip install optimum-cli` to the CI workflow. I'll await #101 before changing the `TextEmbedding` call.

As for the breaking changes, I have the following proposal:

- New module called `output` or similar; which perform two roles:
  1. New `struct` called `EmbeddingOutput`, which is a simple wrapper around `ort::SessionOutputs` generated by `self.session.run(session_inputs)?`. This struct will have methods for exporting different keys and slicing in different ways - I'll have to see how much I can do as `Dim<IxDynImpl>` tends to be pretty annoying.
  2. `embed_raw` or similar method that exports the above, allowing the user to
    - customise the embedding output key
    - recycle the same output for multiple keys without the need to recalculate
  3. `embed` methods in `text_embedding` and `sparse_text_embedding` will now internally use `embed_raw`, and consume the `EmbeddingOutput` instance immediately with preset settings.

Let me know if you have any thoughts regarding the above?

**Comment by Anush008** [comment]
Hey. That's a solid plan. Thank you.

Once we have 3 PRs merged, will do some Q&A tests before releasing v4.

Also, cc @triandco since this plan might be helpful for #99.

**Comment by denwong47** [comment]
I had done some work to the above, only to realise that #99 made some changes to how the array transformation works. Now that `attention_mask` needs to be used during the transformation from `ndarray::Array` to `Vec<Embeddings>`, I need to rethink a little on how to implement `EmbeddingOutputType`; I'll have a think and come back in a few days.

**Comment by denwong47** [comment]
Also in typical Python fashion, installing `optimum` had completely filled up the disk space on Github Runner... I don't know if we should skip that test or give the runner more resources. 😞

**Comment by Anush008** [comment]
I guess we can skip in that case.

**Comment by Anush008** [comment]
This is masterful work @denwong47. Thanks a lot.

**Comment by denwong47** [comment]
> I am a little worried about the many lifetime params and generics, it could push back newer contributors due to the complexity. The library source has been pretty straightforward to grok so far.

I can cut back a bit on that, I think `EmbeddingOutputType` won't be generic to `EmbeddingOutput` anymore, since we need to put in `attention_mask` and the transformation isn't static. So that's one down.

We won't be able to do a lot for the `<'r, 's>` which is from `ort::SessionOutputs` thought; but I'll have a scan again tomorrow to make sure those are well contained within `EmbeddingOutput` and nowhere else.

**Comment by Anush008** [comment]
🙏🙏🙏 

Thanks.

**Comment by denwong47** [comment]
Looking at it, I can also hardcode `f32` as the only type usable with `.try_extract_tensor::<F>()`. This will eliminate all the `F: ort::PrimitiveTensorElementType + Clone` floating around. However this will prevent compatibility with models that use a different floating size; which I gather had not been a requirement so far?

**Comment by Anush008** [comment]
> Which I gather had not been a requirement so far?

Nope. `f32` has sufficed.

File: Cargo.toml
Comment (denwong47) [reply Cargo.toml:42]: As per discussion below, this is not deemed feasible for now.

**Comment by denwong47** [comment]
I have done quite a lot of refactoring to hopefully simplify things a bit. All but one Generics are gone, and I have done some work in preparation for #99, such as a `SingleBatchOutput` class (_please_ help me name this) that allows `attention_mask` to be stored alongside the `session_outputs` so that we can perform per-batch post-processing as we would from that PR.

I probably need to halt work for now until #99 is completed; then I'll merge/replicate the work over.

I have also added some value checking in the tests, in addition to the previous dimensional checks. This is done by summing the vector together and checking the resultant value - this should be safe enough for unit tests.


Let me know any thoughts, much appreciated.

**Review by Anush008 — APPROVED** [review]
Masterful sir @denwong47  🙏

**Comment by denwong47** [comment]
Not sure what's going on with the tests; I'll test on Ubuntu - probably there are minor differences between architecture and OS.

**Comment by Anush008** [comment]
How do you think we should go forward?
Merge this PR and wait for #99 to adapt?
Or should we wait for #99 to be merged and adapt this with the post-process steps it adds?

**Comment by Anush008** [comment]
> probably there are minor differences between architecture and OS?

Maybe try an even lower precision when comparing. 3 digits after . should be good.

**Comment by denwong47** [comment]
I am a bit worried that merging this PR will be a big shock for @triandco and make it very difficult for him to follow. I am happy to for him to work out the pooling first, and I'll adapt mine to suit.

**Comment by Anush008** [comment]
I am leaning towards
> we wait for https://github.com/Anush008/fastembed-rs/pull/99 to be merged and adapt this with the post-process steps it adds?

So that @triandco won't have to bother about the changes here and can continue with the current state of master in #99.

What do you think @denwong47?

**Comment by denwong47** [comment]
> What do you think @denwong47?

Exactly on the same page.

**Comment by Anush008** [comment]
> I am a bit worried that merging this PR will be a big shock for @triandco and make it very difficult for him to follow. I am happy to for him to work out the pooling first, and I'll adapt mine to suit.

Haha. We were on the same lines. Let's do that.

File: src/output/embedding_output.rs
Comment (Anush008) [inline src/output/embedding_output.rs:55]: Hey. Where is this being used?

Comment (denwong47) [reply src/output/embedding_output.rs:55]: Currently its not used, more of a public API than anything. If you want to use the same `EmbeddingOutput` twice, then `consume_with_transformer` is not going to work for you.

I wouldn't mind keeping only one of the two really, and let the use do their own memory management.

Comment (Anush008) [reply src/output/embedding_output.rs:55]: > I wouldn't mind keeping only one of the two really, and let the use do their own memory management.

+1

File: src/output/output_precedence.rs
Comment (Anush008) [inline src/output/output_precedence.rs:50]: Unused?

Comment (denwong47) [reply src/output/output_precedence.rs:50]: I left that in almost as a documentation of existing behaviour, and potentially use it in `sparse_text_embedding`. Mostly because I can't quite understand how these keys work, or why `last_hidden_state` is a chosen special key. I was trying to find documentations but ended up rather more confused than enlightened...

So I imagine a user will be equally baffled by this, and giving them a `DEFAULT` or at least and an example of what to expect will help, hence the detailed explanation of what its doing.

Comment (Anush008) [reply src/output/output_precedence.rs:50]: I think https://github.com/denwong47/fastembed-rs/blob/e0f338508d6fb515c6a3c6c0cc469ae91a44315f/src/text_embedding.rs#L97-L104 is pretty explanatory and will cover for this.

File: src/text_embedding.rs
Comment (Anush008) [inline src/text_embedding.rs:93]: Would suggest keeping this in the separate file.

File: src/output/output_precedence.rs
Comment (denwong47) [reply src/output/output_precedence.rs:50]: Sure, I can remove it.

File: src/text_embedding.rs
Comment (Anush008) [reply src/text_embedding.rs:93]: If you prefer so, sure.

**Comment by denwong47** [comment]
I have rebased and regenerated all the test results using HEAD of #99, everything is passing.

However I feel like something is amiss - if I change the batch size from `None` to `Some(3)` (which forces 2 batches), then the `Mean` tests start failing. The same behaviour is observed in 3.14.1 and the current HEAD, but I thought that was what Pooling was meant to fix? Bottomline is, batching should not have an effect on the embedding values...?

I'll sit down with my Data Scientist tomorrow or Wednesday (@harrysjohnston your time to shine) to go over what's happening - just to make sure we are on the right track here.

Before that, revert this to draft state.

**Comment by denwong47** [comment]
@Anush008 the bottomline is, this PR does not change the output of the embeddings - the test results were generated from using the HEAD of #99 and tests still pass, so I may be trying to deal with an existing issue here, which may be better with a separate PR.

In any case I'll get my dose of education on Pooling from my colleague first, as I could potentially be talking non-sense.

You can advise me on how you want to proceed - merge this first or not.

**Comment by denwong47** [comment]
P.S. I still have to split out `output` as separate file if you want to merge this.

**Comment by Anush008** [comment]
I am willing to merge this to keep the scope limited.

**Comment by denwong47** [comment]
Ok, in that case mod split coming right up.

**Comment by denwong47** [comment]
All done!

**Review by Anush008 — APPROVED** [review]
Thank you sir @denwong47.
Please expect to see this in the v4 release.

**Comment by denwong47** [comment]
Thanks @Anush008 . Great working with you.

I'll report back on the batch and `Pooling::Mean` interaction.

**Comment by Anush008** [comment]
Thank you.

**Pull request merged**
